arangod/GeneralServer/H2CommTask.cpp

```diff
@@ -0,0 +1,654 @@
+  return 0;
+}
+
+template <SocketType T>
+/*static*/ int H2CommTask<T>::on_frame_not_send(nghttp2_session* session,
+                                                const nghttp2_frame* frame,
+                                                int lib_error_code, void* user_data) {
+  if (frame->hd.type != NGHTTP2_HEADERS) {
+    return 0;
+  }
+
+  // Issue RST_STREAM so that stream does not hang around.
+  nghttp2_submit_rst_stream(session, NGHTTP2_FLAG_NONE, frame->hd.stream_id,
+                            NGHTTP2_INTERNAL_ERROR);
+
+  return 0;
+}
+
+template <SocketType T>
+H2CommTask<T>::H2CommTask(GeneralServer& server, ConnectionInfo info,
+                          std::unique_ptr<AsioSocket<T>> so)
+    : GeneralCommTask<T>(server, std::move(info), std::move(so)) {
+  ConnectionStatistics::SET_HTTP(this->_connectionStatistics);
+  initNgHttp2Session();
+}
+
+template <SocketType T>
+H2CommTask<T>::~H2CommTask() noexcept {
+  nghttp2_session_del(_session);
+  _session = nullptr;
+  if (!_streams.empty()) {
+    LOG_TOPIC("924cf", WARN, Logger::REQUESTS) << "got "
+      << _streams.size() << " remaining H2 streams";
+  }
+}
+
+/// init h2 session
+template <SocketType T>
+void H2CommTask<T>::initNgHttp2Session() {
+  nghttp2_session_callbacks* callbacks;
+  int rv = nghttp2_session_callbacks_new(&callbacks);
+  if (rv != 0) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+    return;
+  }
+
+  auto cbScope = scopeGuard([&] { nghttp2_session_callbacks_del(callbacks); });
+
+  nghttp2_session_callbacks_set_on_begin_headers_callback(callbacks, H2CommTask<T>::on_begin_headers);
+  nghttp2_session_callbacks_set_on_header_callback(callbacks, H2CommTask<T>::on_header);
+  nghttp2_session_callbacks_set_on_frame_recv_callback(callbacks, H2CommTask<T>::on_frame_recv);
+  nghttp2_session_callbacks_set_on_data_chunk_recv_callback(callbacks, H2CommTask<T>::on_data_chunk_recv);
+  nghttp2_session_callbacks_set_on_stream_close_callback(callbacks, H2CommTask<T>::on_stream_close);
+  nghttp2_session_callbacks_set_on_frame_send_callback(callbacks, H2CommTask<T>::on_frame_send);
+  nghttp2_session_callbacks_set_on_frame_not_send_callback(callbacks, H2CommTask<T>::on_frame_not_send);
+
+  rv = nghttp2_session_server_new(&_session, callbacks, /*args*/ this);
+  if (rv != 0) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+    return;
+  }
+}
+
+template <SocketType T>
+void H2CommTask<T>::upgrade(std::unique_ptr<HttpRequest> req) {
+  bool found;
+  std::string const& settings = req->header("http2-settings", found);
+  bool const wasHead = req->requestType() == RequestType::HEAD;
+
+  const uint8_t* src = reinterpret_cast<const uint8_t*>(settings.data());
+  int rv = nghttp2_session_upgrade2(_session, src, settings.size(), wasHead, nullptr);
+
+  if (rv != 0) {
+    if (rv == NGHTTP2_ERR_INVALID_ARGUMENT) {
+      // The settings_payload is badly formed.
+      LOG_TOPIC("0103c", ERR, Logger::REQUESTS) << "invalid upgrade header";
```

Comment:
I guess we don't want loglevel `ERR` here, because it is an application (client-side) error, and not some problem in arangod.

```suggestion
      LOG_TOPIC("0103c", INFO, Logger::REQUESTS) << "invalid upgrade header";
```